remove qemu's systick workaround

agansari opened the pull request:
Removed workarounds in systick driver as they prevent normal usage in TICKLESS systems. 
Set qemu parameter for MPS2 targets: -icount auto.
CONFIG_SOC_SERIES_MPS2 specific test workaround is removed to use systick drives as tickless.
test_sched_timeslice_reset - loosen tolerance for slow ticking systems.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari commented:
@galak It's the opposite way around, both my MCU had 48MHz clocks, and CONFIG_SYS_CLOCK_TICKS_PER_SEC=10.000 as default tickless this means a CYC_PER_TICK=4800. Overflow condition was hardly reached and cycle count was always incremented regardless of overflow.

andyross reviewed (approved):
This looks safe and correct (one style/optimization nitpick).  I'm confused why you were seeing early interrupts though.  Under what circumstances will a CPU throw a timer interrupt if the COUNTFLAG bit is **not** set?!  Per documentation that should never happen, as I read it.

Is something else hooked to your timer interrupt on the affected platorms maybe?

agansari commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -87,14 +87,16 @@ void z_clock_isr(void *arg)
+	u32_t bool_cnt_flag;
 
-	cycle_count += last_load;
+	bool_cnt_flag = (SysTick->CTRL & SysTick_CTRL_COUNTFLAG_Msk); /* Reset overflow flag */
```
@pabigot i find #15487 as being a better solution to this issue, but as far as i know it has some issues on other boards, maybe that pull should be revisited @ioannisg

agansari commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -87,14 +87,16 @@ void z_clock_isr(void *arg)
+	bool_cnt_flag >>= SysTick_CTRL_COUNTFLAG_Pos;
+	overflow_cyc = 0;
+
+	cycle_count += last_load * bool_cnt_flag;
```
Yes I've wrote it this way to avoid a jump condition, but it has no advantage in thumb code.

Will change it to a?b:c as in other places in this file.

ioannisg commented:
We really need to do something with this - the fifo-timeout test is now failing for several ARM boards with Systick timer

agansari commented:
> We really need to do something with this - the fifo-timeout test is now failing for several ARM boards with Systick timer

@ioannisg @galak @pabigot @andyross I've resolved part of the newer issues that it introduces, it starts to look a lot like #15487 and i think that should rather be merged as it's better comented.
The issue i'm still having is `tests/kernel/sched/schedule_api ` failing on MPS2 qemu.

agansari commented:
@ioannisg @galak @pabigot @andyross fixed the timer's issues in qemu, patch should be good now.

ioannisg commented:
> 
> 
> @ioannisg @galak @pabigot @andyross fixed the timer's issues in qemu, patch should be good now.

Let's add a HW Block - and I am going to test this on whatever HW I have.
I think we need to be confident that this works on as many platforms as possible

agansari commented:
Thank you @ioannisg

ioannisg commented:
`tests/kernel/timer` (`timer_api `and `timer_monotonic`) are failing for `atsamr21_xpro`. The tests are passing on `sam_e70_xplained` board.

ioannisg commented:
`tests/kernel/sched` is failing for `lpcxpresso54114_m4`.

ioannisg commented:
All tests are passing on `frdm_k64f`

agansari commented:
> `tests/kernel/sched` is failing for `lpcxpresso54114_m4`.

@ioannisg this is #18282, i have also observed this issue showed-up again. It's somewhat inconsistent, if i change test order, or add debug prints it disappears. Think we can reopen #18282 an continue debug from there.

pabigot commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -88,13 +88,18 @@ void z_clock_isr(void *arg)
+	/* QEMU will send IRQ, but not sure about setting SysTick->CTRL */
 	cycle_count += last_load;
+#endif
+	overflow_cyc = 0;
```
This isn't a new problem, but I'm concerned about a race condition here: this mutates `overflow_cyc` without holding the spinlock.  If a higher priority interrupt invokes `elapsed()` through any other function (like `z_clock_elapsed()` or `z_timer_cycle_get_32()` or `z_clock_set_timeout()`) information may be lost.

agansari replied:
@pabigot you are right, as all other uses of `elapsed()` are guarded by spinlocks, i've added an irq lock so no other ISR would also call `elapsed()`. 
I was debugging #18282 and what @andyross suggested (blocking multiple irq interrupts by using a busy wait). Looks like guarding the variable changes in `elapsed()` makes the behavior coherent.

pabigot reviewed (dismissed):
Noting what looks like a race condition.  If it isn't I'll switch to "approve" so we can get rc3 out.  @ioannisg

pabigot commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -87,15 +87,17 @@ void z_clock_isr(void *arg)
-
 	z_clock_announce(TICKLESS ? dticks : 1);
+
+	irq_unlock(key);
```
I'm very certain you do **not** want the lock held during `z_clock_announce()` which can take arbitrarily long.   Move it before that call, please.

agansari replied:
@pabigot thanks! pushed again.

pabigot reviewed (approved):
I don't have a hardware platform to test this, but by inspection it makes sense.

ioannisg commented:
@agansari tests/kernel/timer/timer_api fails on frdm_k64f, AFAICS - can you check?

ioannisg commented:
Fails also on sam_e70_xplained and atsamr21_xpro

ioannisg commented:
I moved this to 2.1 since some test issues were not resolved in time for 2.0

pabigot commented:
@ioannisg are the failures on real hardware?  tests/kernel/timer/timer_api passes on frdm_k64f for me with both default and tickless configurations.

I also looked at a couple of the shippable failures on mps2_an385 yesterday and they seem likely to be due to qemu per https://github.com/zephyrproject-rtos/zephyr/pull/18651#discussion_r319100694; they also worked on k64f hardware.  I don't have an385 hardware to check.

ioannisg commented:
> 
> 
> @ioannisg are the failures on real hardware? tests/kernel/timer/timer_api passes on frdm_k64f for me with both default and tickless configurations.
> 
> I also looked at a couple of the shippable failures on mps2_an385 yesterday and they seem likely to be due to qemu per [#18651 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/18651#discussion_r319100694); they also worked on k64f hardware. I don't have an385 hardware to check.

@pabigot thanks really for looking at this. So, it seems that @agansari also reports the test passing on frdm_k64f. But I see it constantly failing on mine. And also on the ATSAM boards I tested it.

@galak @nashif could you do some last minute testing here? E.g. try on a couple of SysTick ARM boards. Otherwise, I would say it's better to finalize this after 2.0 release.

agansari committed:
drivers: remove arm's qemu systick workaround

Removed workarounds in systick driver as they prevent normal usage in
TICKLESS systems. Driver still behaved like an interrupt based ticker.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
boards: MPS2 targets enabled icount

Set qemu parameter for MPS2 targets: -icount auto

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
tests: scheduler_api removed MPS2 workaround

CONFIG_SOC_SERIES_MPS2 specific test workaround is removed
to use systick drives as tickless.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
tests: schedule_api adapt to slow ticks

This test case is has a tolerance of 1ms, but systems with a tick slower
than 1000 ticks/sec may spil outside the 1ms tolerance.
Tolerance adapts to system's ticks/sec, e.g. QEMU targets have
100ticks/sec -> tolerance is 10ms.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari commented:
Issue is still under debugging, there are sanity checks that still fail randomly: sched multiq and tickless concept on MPS2 qemu.
Can't test on atsam boards.

wentongwu commented:
oh, just see you are enabling Qemu icount, Also I'm doing this work(see https://github.com/wentongwu/zephyr/tree/qemu and https://github.com/wentongwu/zephyr/tree/icount), I'm running stress test with "for i in 
 {1.. 100000}; do ./scripts/sanitycheck -p mps2_an385 -T tests/kernel/ --icount-shift 5 -v ; done" with branch https://github.com/wentongwu/zephyr/tree/icount

agansari commented:
@wentongwu yes, i am aware of #14173 which is trying to remove icount. I had to use icount as qemu was unpredictable without it; the way systick was implemented in qemu was a workaround.

wentongwu commented:
> @wentongwu yes, i am aware of #14173 which is trying to remove icount. I had to use icount as qemu was unpredictable without it; the way systick was implemented in qemu was a workaround.

Why you say “ #14173 which is trying to remove icount“?

agansari commented:
> Why you say “ #14173 which is trying to remove icount“?

Sorry, wanted to say remove the workaround and enable icount

wentongwu commented:
ok, Thanks. 
"the way systick was implemented in qemu was a workaround" could you please talk more about this? Actually I debug the systick in Qemu(armv7m_systick.c) with gdb for icount mode, at least for software emulation, I think it works. @agansari Thanks.

agansari commented:
@wentongwu i have a problem with the following line of code:
`z_clock_announce(TICKLESS ? dticks : 1);`
with the workaround it always behaves as a tick based system, not a tickless one where it counts how far it got, it relies on interrupts and not the actual timer value. 
This is why i removed `CONFIG_QEMU_TICKLESS_WORKAROUND` from `TICKLESS `macro

wentongwu commented:
> @wentongwu i have a problem with the following line of code:
> `z_clock_announce(TICKLESS ? dticks : 1);`
> with the workaround it always behaves as a tick based system, not a tickless one where it counts how far it got, it relies on interrupts and not the actual timer value.
> This is why i removed `CONFIG_QEMU_TICKLESS_WORKAROUND` from `TICKLESS `macro

Ok, Thanks, I will consider it. 
And I thought you found some bugs in Qemu for sysTick.

carlescufi commented:
@galak please merge this if ready

ioannisg commented:
> 
> 
> @galak please merge this if ready

@galak @carlescufi I don't this this is ready.
@agansari pls, advice.

galak commented:
> @galak @carlescufi I don't this this is ready. Have we reviewed the tolerance introduced in the test?
> @agansari please, advice. :)

Need to re-test, not sure we've addressed the failing testcases that this caused.

agansari commented:
@galak @carlescufi @ioannisg i am still investigating on MPS2 QEMU sched and tickless concept tests, so it's not ready to merge.
Can we have someone from atmel look at why it's failing on sam_e70_xplained/atsamr21_xpro? I don't have these boards.

pabigot reviewed (changes requested):
I've compared this solution against #15487.

* The original problem with tests/kernel/fifo/fifo_timeout/prj_poll was reproduced on atsamd21_xpro on master d5633bc395b2bc8ce9f7685e8df9ead43deb20c3: it always fails
* I rebased this on master d5633bc395b2bc8ce9f7685e8df9ead43deb20c3
* tests/kernel/fifo/fifo_timeout/prj_poll passes reliably on atsamd21_xpro

However this PR incorporates other changes, one of which looks very wrong, and I'm still uncomfortable with the way this deals with the systick issue (see https://github.com/zephyrproject-rtos/zephyr/pull/18651#discussion_r317940005).

I propose that #15487 be re-reviewed and if no other concerns are discovered that it be merged.  The additional patches in this PR can then be checked to see if they're still relevant, and merged on their own merits.

agansari commented on tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c:
```diff
@@ -22,6 +22,15 @@ BUILD_ASSERT(NUM_THREAD <= MAX_NUM_THREAD);
 	((u64_t)(HALF_SLICE_SIZE) * sys_clock_hw_cycles_per_sec() / 1000)
 
+/* Task switch tolerance ... */
+#if CONFIG_SYS_CLOCK_TICKS_PER_SEC < 1000
```
@pabigot you are right, i reversed the condition.

agansari commented:
@pabigot @ioannisg I agree to merging #15487 and then fixing the other issues related to QEMU and slow ticks/sec in this pull.

ioannisg commented:
> 
> 
> @pabigot @ioannisg I agree to merging #15487 and then fixing the other issues related to QEMU and slow ticks/sec in this pull.

We cannot merge #15487 with QEMU failures; we need to resolve them somehow, or exclude the platforms from these tests. Ideally, we should find out if we can improve QEMU boards so the tests do not fail, or go ahead with your proposal of adding some slack in the error-prone tests, e.g. the scheduler

pabigot commented on tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c:
```diff
@@ -19,8 +19,17 @@ BUILD_ASSERT(NUM_THREAD <= MAX_NUM_THREAD);
+#define TASK_SWITCH_TOLERANCE (1)
+#else
+/* ... 1ms is faster than a tick, loosen tolerance to 1 tick */
+#define TASK_SWITCH_TOLERANCE (1000 / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
```
This is vulnerable to a repeated-truncation problem that's showed up multiple times in the past.  For example, if `CONFIG_SYS_CLOCK_TICKS_PER_SEC` is 256 the tolerance should be 3.90625 ms, but it'll be truncated to 3 ms.  In actual use durations of 256 ticks will appear to be either 3 or 4 ms long, averaging to the true value.

Below the tolerance is used to calculate both a lower bound and an upper bound, where it is again truncated.  The problem is that for the lower bound it's appropriate to subtract 3, but for the upper bound you must add 4.  If you only add 3 you get intermittent failures in the cases where the experienced duration maps to 4 ms.

This is exactly the sort of thing that #19591 will help with, by providing conversion functions that round in the direction necessary for the specific use, so in hopes that we get that unstuck and merged soon I'm ok with merging this as is, with a plan to replace the calculations with accurate ones once infrastructure is available to support that.

agansari replied:
Thank you @pabigot and @ioannisg.
I changed the title and description to reflect the new purpose of this PR.

Should we have this PR merged and then add a new PR for mathematical rounding or would something like `ceiling_fraction` macro do the job in a new force push? @pabigot

pabigot replied:
Yes, I'd like this to go in as-is.  We're waiting on #19591 to provide `k_ticks_to_ms_ceil32()` API to help with the potential concern I identified.

pabigot reviewed (approved):
An observation below, I don't think it should block this.

#15847 rebased on top of this passes all mps2_an385 tests/kernel checks\*.  Without this PR under it two fail consistently.

\* timer_api/kernel.timer  still fails intermittently but passes on a second try, unlike kernel/sched/schedule_api/kernel.sched.* which persistently fail without this.

ioannisg reviewed (approved):
This PR looks OK to me. I am happy that the workaround for MPS2 is removed. I ran all tests locally for mps2_an385 and mps2_an521; they all pass.

What is left to think about is whether more timing-sensitive tests require this kind of tolerance. I feel it is not only the sched tests; at least the timer_api and several benchmark tests are good candidates.

ioannisg commented:
@andyross would you be able to take a look at the tolerance added to the kernel/sched tests?

Other than that, @nashif @MaureenHelm,  this is an ARM & ARM QEMU -only change, and looks merge-able to me
